discretisedfield/mesh.py

```diff
@@ -386,15 +389,51 @@ def axis_points(self, axis, /):
         >>> cell = (2, 1, 1)
         >>> mesh = df.Mesh(region=df.Region(p1=p1, p2=p2), cell=cell)
         ...
-        >>> list(mesh.axis_points('x'))
+        >>> list(mesh.midpoints.x)
         [1.0, 3.0, 5.0, 7.0, 9.0]
 
         """
-        if isinstance(axis, str):
-            axis = dfu.axesdict[axis]
+        midpoints = collections.namedtuple('midpoints', ['x', 'y', 'z'])
+
+        return midpoints(*(
+            np.linspace(pmin + cell / 2, pmax - cell / 2, n)
+            for pmin, pmax, cell, n in
+            zip(self.region.pmin, self.region.pmax, self.cell, self.n)
+        ))
+
+    @functools.cached_property
+    def vertices(self):
+        """Vertices of the cells of the mesh along the three directions.
+
+        This method returns a named tuple with generators yielding vertices of
+        the cells along the three spatial directions. Individual directions can
+        be accessed from the tuple.
+
+        Returns
+        -------
+        collections.namedtuple
+
+            Generators for vertices of the cells along ``x``, ``y``, and ``z``.
+
+        Examples
+        --------
+        1. Getting vertices along the ``x`` axis.
+
+        >>> import discretisedfield as df
+        ...
+        >>> p1 = (0, 0, 0)
+        >>> p2 = (10, 1, 1)
+        >>> cell = (2, 1, 1)
+        >>> mesh = df.Mesh(region=df.Region(p1=p1, p2=p2), cell=cell)
+        ...
+        >>> list(mesh.vertices.x)
+        [0, 2, 4, 6, 8, 10]
+
+        """
+        vertices = collections.namedtuple('vertices', ['x', 'y', 'z'])
 
-        for i in range(self.n[axis]):
-            yield self.index2point((0, 0, 0))[axis] + i*self.cell[axis]
+        return vertices(*(np.linspace(pmin, pmax, n + 1) for pmin, pmax, n in
+                          zip(self.region.pmin, self.region.pmax, self.n)))
```

Comment:
@marijanbeg If you are happy with this (not anymore lazy) calculation of the `vertices`/`midpoints` we can merge the PR.

Reply:
[Copied from Zulip, looks like you've missed it there, and slightly extended]
I think using `np.linspace` would be a lot better, because:
- it is generally more convenient to use (no need to convert to a list first) and all parts within `discretisedfield` can use/need a numpy array "version" anyways.
- faster for larger n (example: 2000 cells, generator: 750μs, numpy: 30 μs (raw numpy), 200 μs with the additional calls to the mesh/region); subsequent calls are ~50 ns (due to caching).
- easier to use because we can assign and reuse it, whereas more care is necessary with the generator expression.